### PR TITLE
feat: filter students by ref and names

### PR DESCRIPTION
**Pull request by hei-ryan**

### CODE REVIEW

**Comment by hei-ryan** [comment]
Actually, the pagination with filters doesn't work so the test should fail but you can already comments the filters test.

File: src/main/java/school/hei/haapi/endpoint/rest/controller/StudentController.java [hunk 1]
```diff
@@ -24,34 +24,37 @@
...
 @AllArgsConstructor
 public class StudentController {
-
   private final UserService userService;
   private final UserMapper userMapper;
 
   @GetMapping("/students/{id}")
   public Student getStudentById(
-      @AuthenticationPrincipal Principal principal, @PathVariable String id) {
+          @AuthenticationPrincipal Principal principal, @PathVariable String id) {
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/controller/StudentController.java]: non

File: src/main/java/school/hei/haapi/endpoint/rest/controller/StudentController.java [hunk 2]
```diff
@@ -24,34 +24,37 @@
...
     if (Role.STUDENT.getRole().equals(principal.getRole())
-        && !id.equals(principal.getUser().getId())) {
+            && !id.equals(principal.getUser().getId())) {
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/controller/StudentController.java]: no

File: src/main/java/school/hei/haapi/endpoint/rest/controller/StudentController.java [hunk 3]
```diff
@@ -24,34 +24,37 @@
...
   }
 
   @PutMapping("/students")
   public List<Student> saveAll(@RequestBody List<Student> toWrite) {
     return userService
-        .saveAll(toWrite.stream().map(userMapper::toDomain).collect(toUnmodifiableList()))
-        .stream()
-        .map(userMapper::toRestStudent)
-        .collect(toUnmodifiableList());
+            .saveAll(toWrite.stream().map(userMapper::toDomain).collect(toUnmodifiableList()))
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/controller/StudentController.java]: no

File: src/main/java/school/hei/haapi/endpoint/rest/controller/StudentController.java [hunk 4]
```diff
@@ -24,34 +24,37 @@
...
       throw new ForbiddenException();
     }
     return userMapper.toRestStudent(userService.getById(id));
   }
 
   @GetMapping("/students")
   public List<Student> getStudents(
-      @RequestParam PageFromOne page, @RequestParam("page_size") BoundedPageSize pageSize) {
-    return userService.getByRole(User.Role.STUDENT, page, pageSize).stream()
+          @RequestParam PageFromOne page, @RequestParam("page_size") BoundedPageSize pageSize,
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/controller/StudentController.java]: no

File: src/main/java/school/hei/haapi/endpoint/rest/controller/StudentController.java [hunk 5]
```diff
@@ -24,34 +24,37 @@
...
+          @RequestParam(value = "ref", required = false) String ref,
+          @RequestParam(value = "first_name", required = false) String firstName,
+          @RequestParam(value = "last_name", required = false) String lastName) {
+    return userService.getByCriteria(User.Role.STUDENT, page, pageSize,
+                    ref, firstName, lastName).stream()
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/endpoint/rest/controller/StudentController.java]: too much indentation

File: src/main/java/school/hei/haapi/service/UserService.java [hunk 6]
```diff
@@ -37,24 +46,51 @@ public User getByEmail(String email) {
...
-        Sort.by(ASC, "ref"));
+            page.getValue() - 1,
+            pageSize.getValue(),
+            Sort.by(ASC, "ref"));
     return userRepository.getByRole(role, pageable);
   }
+
+  public List<User> getByCriteria(User.Role role, PageFromOne page, BoundedPageSize pageSize,
+                                  String ref, String firstName, String lastName) {
+    CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
```
Comment (hei-teacher) [inline src/main/java/school/hei/haapi/service/UserService.java]: No. Create UserRepository::findByReferenceAndFirstNameAndLastName(ref, first, last).
Equality is `Like`. Default is `like %` for everything.

File: src/test/java/school/hei/haapi/integration/StudentIT.java [hunk 7]
```diff
@@ -170,40 +189,106 @@ void teacher_read_ok() throws ApiException {
...
+  void manager_read_by_ref_and_name_ok() throws ApiException {
+    ApiClient manager1Client = anApiClient(MANAGER1_TOKEN);
+    UsersApi api = new UsersApi(manager1Client);
+
+    List<Student> actualStudents = api.getStudents(1, 20, student1().getRef(),
+            student1().getFirstName(), student1().getLastName());
+
+    assertEquals(1, actualStudents.size());
+    assertTrue(actualStudents.contains(student1()));
+    assertFalse(actualStudents.contains(student2()));
```
Comment (hei-teacher) [inline src/test/java/school/hei/haapi/integration/StudentIT.java]: Useless

File: src/test/java/school/hei/haapi/integration/StudentIT.java [hunk 8]
```diff
@@ -170,40 +189,106 @@ void teacher_read_ok() throws ApiException {
...
+  @Test
+  void manager_read_by_ref_and_bad_name_ko() throws ApiException {
+    ApiClient manager1Client = anApiClient(MANAGER1_TOKEN);
+    UsersApi api = new UsersApi(manager1Client);
+
+    List<Student> actualStudents = api.getStudents(1, 20, student2().getRef(),
+            null, student1().getLastName());
+
+    assertEquals(0, actualStudents.size());
+    assertFalse(actualStudents.contains(student1()));
```
Comment (hei-teacher) [inline src/test/java/school/hei/haapi/integration/StudentIT.java]: why...

File: src/test/java/school/hei/haapi/integration/StudentIT.java [hunk 9]
```diff
@@ -170,40 +189,106 @@ void teacher_read_ok() throws ApiException {
...
+    assertFalse(actualStudents.contains(student2()));
```
Comment (hei-teacher) [inline src/test/java/school/hei/haapi/integration/StudentIT.java]: why...

File: src/test/java/school/hei/haapi/integration/StudentIT.java [hunk 10]
```diff
@@ -170,40 +189,106 @@ void teacher_read_ok() throws ApiException {
...
+    assertFalse(actualStudents.contains(student3()));
```
Comment (hei-teacher) [inline src/test/java/school/hei/haapi/integration/StudentIT.java]: why...

File: src/test/java/school/hei/haapi/integration/TeacherIT.java [hunk 11]
```diff
@@ -138,11 +157,76 @@ void teacher_read_ko() {
...
+            "{\"type\":\"403 FORBIDDEN\",\"message\":\"Access is denied\"}",
+            () -> api.getTeacherById(TEACHER2_ID));
     assertThrowsApiException(
-        "{\"type\":\"403 FORBIDDEN\",\"message\":\"Access is denied\"}",
-        () -> api.getTeachers(1, 20));
+            "{\"type\":\"403 FORBIDDEN\",\"message\":\"Access is denied\"}",
+            () -> api.getTeachers(1, 20, null, null, null));
+  }
+
+  @Test
```
Comment (hei-teacher) [inline src/test/java/school/hei/haapi/integration/TeacherIT.java]: No. Use intergration test for just teacher for example, but use unit tests for the rest.

**Review by hei-teacher — CHANGES_REQUESTED** [review]

File: src/test/java/school/hei/haapi/integration/TeacherIT.java (hunk 11 above)
Comment (hei-ryan) [reply src/test/java/school/hei/haapi/integration/TeacherIT.java]: After thinking a while, is it really necessary to add UT for teacher controller if i just added default values into @RequestParams and the service, already tested in StudentIT, is doing the rest.

**Review by hei-teacher — APPROVED** [review]

**Pull request merged**